Outline of new tutorial for UBI in AWS Managed Services (#10737)
opensearch-project/documentation-website: _search-plugins/ubi/ubi-aws-managed-services-tutorial.md

---
layout: default
title: Using UBI in Amazon OpenSearch Service
parent: User Behavior Insights
has_children: false
nav_order: 30
---

# Using UBI in Amazon OpenSearch Service

> This tutorial shows you how to collect queries and events in the User Behavior Insights (UBI) format when using Amazon OpenSearch Service. After following this tutorial, you’ll be able to send authenticated queries and events to both Amazon S3 for long-term storage and OpenSearch for real-time processing using the `curl` command-line tool. 

Reviewer: ```suggestion
This tutorial shows you how to collect queries and events in the User Behavior Insights (UBI) format when using Amazon OpenSearch Service. After following this tutorial, you'll be able to send authenticated queries and events to both Amazon Simple Storage Service (Amazon S3) for long-term storage and OpenSearch for real-time processing using the `curl` command-line tool. 
```

This tutorial assumes the following:

1. You are using Amazon OpenSearch Service version 2.19.
2. You are not using the UBI plugin for OpenSearch, which becomes available in version 3.1 for Amazon OpenSearch Service.

> 3. You are writing UBI data to OpenSearch using [OpenSearch Ingestion](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/ingestion.html), the managed version of Data Prepper.

Reviewer: ```suggestion
3. You are writing UBI data to OpenSearch using [Amazon OpenSearch Ingestion](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/ingestion.html), the managed version of OpenSearch Data Prepper.
```

> 4. You have already configured permissions between OpenSearch Ingestion and your managed clusters by completing the [Tutorial: Ingesting data into a domain using Amazon OpenSearch Ingestion](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/osis-get-started.html), specifically the *Required permissions* step.

Reviewer: ```suggestion
4. You have already configured permissions between OpenSearch Ingestion and your managed clusters by following the instructions found in [Tutorial: Ingesting data into a domain using Amazon OpenSearch Ingestion](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/osis-get-started.html), specifically the *Required permissions* step.
```

## Step 1: Set up OpenSearch indexes for UBI

Follow these steps to create the indexes needed for UBI data:

1. Log in to OpenSearch Dashboards in Amazon OpenSearch Service.
1. On the main menu, select **Management > Dev Tools** to open the **Dev Tools** console.
1. Create two new indexes: `ubi_events` and `ubi_queries`.

    1. First, start creating the mappings for the `ubi_events` index:

        ```json
        PUT /ubi_events
        {
          "mappings": 
        }
        ```

>         A syntax warning will appear at this point. That’s expected; you’ll enter the mappings next.

Reviewer: ```suggestion
        A syntax warning will appear at this point. That's expected; you'll enter the mappings next.
```

        Open the [events-mapping.json](https://github.com/opensearch-project/user-behavior-insights/blob/main/src/main/resources/events-mapping.json) file, copy its contents, and paste them after the `"mappings":` line:

        ```json
        PUT ubi_events
        {
          "mappings": {
            "properties": {
              "application": {
                "type": "keyword",
                "ignore_above": 256
              },
              "action_name": {
                "type": "keyword",
                "ignore_above": 100
              },
              ...
            }
          }
        }
        ```

        Run the command and verify that it succeeds.

    1. Next, create the `ubi_queries` index in a similar way. Set up the mappings:

        ```json
        PUT ubi_queries
        {
          "mappings": 
        }
        ```

        Open the [queries-mapping.json](https://github.com/opensearch-project/user-behavior-insights/blob/main/src/main/resources/queries-mapping.json?utm_source=chatgpt.com) file, copy its contents, and paste them after the `"mappings":` line. Run the command and verify that it succeeds.

## Step 2: Set up Amazon S3 storage

> For long-term storage of UBI data, use Amazon Simple Storage Service (Amazon S3).  

Reviewer: ```suggestion
For long-term storage of UBI data, use Amazon S3.  
```

> Before proceeding, create an S3 bucket in which the queries and events data will be stored. You can do this in the AWS Management Console. Note the bucket name and the region where it is created; you’ll need this information for the following steps.

Reviewer: ```suggestion
Before proceeding, create an S3 bucket in which the query and event data will be stored. You can do this in the AWS Management Console. Note the bucket name and the AWS Region in which it is created; you'll need this information for the following steps.
```

{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "es:ESHttpPost",
                "es:ESHttpPut",
                "es:ESHttpGet",
                "es:ESHttpHead"
            ],
            "Resource": "arn:aws:es:*:<YOUR_AWS_ACCOUNT_NUMBER>:*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "s3:PutObject"
            ],
            "Resource": "arn:aws:s3:::<YOUR-BUCKET>/*"
        }
    ]
}
```
{% include copy.html %}

### Step 3(a): Create a query pipeline

Follow these steps to create a pipeline for UBI query data:

1. In the Amazon OpenSearch Service console, select **Pipelines** from the left navigation pane.
1. Select **Create pipeline**.
1. Select a **Blank** pipeline, then select **Select blueprint**.
1. Configure the pipeline to use the **HTTP** source plugin, which accepts UBI query data in JSON array format. Set the OpenSearch Service domain as the sink, directing all data into the `ubi_queries` index. Additionally, log all events to an S3 bucket in `.ndjson` format.
1. In the **Source** menu, select **HTTP**. For **Path**, enter `/ubi/queries`.

> 1. For **Source network options**, select **Public access** to allow posting data from your application.

Reviewer: ```suggestion
1. For **Source network options**, select **Public access** to allow the posting of data from your application.
```

1. Select **Next**.
1. Skip intermediate **Processor** steps by selecting **Next** on the **Processor** screen.
1. Configure the first sink:
   * In **OpenSearch resource type**, select **Managed cluster**.  
   * Select the OpenSearch Service domain you created earlier.  
   * In **Index name**, enter `ubi_queries`. Make sure this index exists with the required UBI schema.
1. Configure the second sink:
    * Select **Add Sink**.  
    * Select **Amazon S3**.  

>     * Enter the bucket name and region you created previously.  

Reviewer: ```suggestion
    * Enter the bucket name and AWS Region you created previously.  
```

    * In **Event Collection Timeout**, enter `60s` to observe data flow quickly.  
    * Select **NDJSON** as the format.
1. Select **Next**.
1. Name the pipeline `ubi-queries-pipeline` and leave the capacity settings at their defaults.
1. Select **Next**, then **Create Pipeline**.

### Step 3(b): Test the query pipeline

> When the pipeline status is `Active`, you can start ingesting data into it. You must sign all HTTP requests to the pipeline using [Signature Version 4](https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html). Use an HTTP tool such as [Postman](https://www.getpostman.com/) or [awscurl](https://github.com/okigan/awscurl) to send some data to the pipeline. As with indexing data directly to a domain, ingesting data into a pipeline always requires either an IAM role or an [IAM access key and secret key](https://docs.aws.amazon.com/powershell/latest/userguide/pstools-appendix-sign-up.html).

Reviewer: ```suggestion
When the pipeline status is `Active`, you can start ingesting data into it. You must sign all HTTP requests to the pipeline using [AWS Signature Version 4](https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html). Use an HTTP tool such as [Postman](https://www.getpostman.com/) or [awscurl](https://github.com/okigan/awscurl) to send some data to the pipeline. As with indexing data directly to a domain, ingesting data into a pipeline always requires either an AWS Identity and Access Management (IAM) role or an [IAM access key and secret key](https://docs.aws.amazon.com/powershell/latest/userguide/pstools-appendix-sign-up.html).
```

To test the pipeline, use these steps:

1. Retrieve the ingestion URL from the **Pipeline settings** page, shown in the following image.

    ![Pipeline Settings]({{site.url}}{{site.baseurl}}/images/ubi/opensearch-ingestion-pipeline.png "Pipeline Settings")

> 1. Post a UBI query to the ingestion pipeline. The following is an example of posting a query using [awscurl](https://github.com/okigan/awscurl):

Reviewer: ```suggestion
1. Post a UBI query to the ingest pipeline. The following is an example of posting a query using [awscurl](https://github.com/okigan/awscurl):
```
